src/typeconversions.jl

```diff
@@ -60,6 +60,25 @@ function datatype(str::VLen{C}) where {C<:CharType}
     Datatype(API.h5t_vlen_create(type_id))
 end
 
+# Compound types
+
+datatype(::T) where {T} = Datatype(hdf5_type_id(T), false)
```

Comment:
why `toclose = false`?

Reply:
To be honest, I am not sure why this is needed, I took it over from @mkitti

Reply:
I'm just following similar methods here:
https://github.com/JuliaIO/HDF5.jl/blob/987d9c89e43abe68dfeced963314ed09ad4c1803/src/typeconversions.jl#L197-L199

Reply:
We're also defining new `hdf5_type_id` below.

Using the finalizer can get messy due to multithreading issues. What we should consider, perhaps in another pull request, is a `do` method where we automatically close the `Datatype` in the current thread.

Reply:
> We're also defining new `hdf5_type_id` below.

What I meant is that e,g, `hdf5_type_id(::Type{Int64})` is just accessing a constant, it doesn't create a new handle. The new method in this PR (https://github.com/JuliaIO/HDF5.jl/pull/1013/files#diff-bf557f1b887a640c4fe55cee7279c9f621001ff726a876b9f7bfbd7f8a7d0b44R70) calls `h5t_create`, creating a new handle. They need to be cleaned up, or you'll have a memory leak.

I'm not sure I understand the threading issue: we have finalizers on lots of other objects, why would Datatypes be special in this case?

Reply:
The finalizers are problematic elsewhere as well. But we also have a `do` interface in many of those cases which helps.

Quoting Jeff Benzanson in https://github.com/JuliaLang/julia/issues/11207 :

> Finalizers are inefficient and unpredictable. And with the new GC, it might take much longer to get around to freeing an object, therefore tying up its resources longer. Ideally releasing external resources should not be tied to how memory management works.

The problem is particularly acute here since the underlying HDF5 C library is not thread-safe. https://github.com/JuliaLang/julia/pull/45272 improves the situation by making the finalizer calls eager and deterministic, making it likely that the finalizer calls `H5T_close` using the same thread.

The better solution here is to either explicitly call `close` which should invoke `HDF5.API.h5t_close` or we provide a `do` syntax to assist with this.

Reply:
I don't disagree, and adding some sort of `do` syntax pattern is a good idea, but that doesn't preclude having finalizers as well for when that approach isn't feasible.

Within the scope of this PR, these objects should be finalized, or otherwise you're going to get memory leaks.